pkg/controller/egress/controller.go

```diff
@@ -239,32 +241,34 @@ func (c *EgressController) setIPAllocation(egressName string, ip net.IP, poolNam
 }
 
 // syncEgressIP is responsible for releasing stale EgressIP and allocating new EgressIP for an Egress if applicable.
-func (c *EgressController) syncEgressIP(egress *egressv1beta1.Egress) (net.IP, error) {
+func (c *EgressController) syncEgressIP(egress *egressv1beta1.Egress) (net.IP, *egressv1beta1.Egress, error) {
 	prevIP, prevIPPool, exists := c.getIPAllocation(egress.Name)
 	if exists {
 		// The EgressIP and the ExternalIPPool don't change, do nothing.
 		if prevIP.String() == egress.Spec.EgressIP && prevIPPool == egress.Spec.ExternalIPPool && c.externalIPAllocator.IPPoolExists(egress.Spec.ExternalIPPool) {
-			return prevIP, nil
+			return prevIP, egress, nil
 		}
 		// Either EgressIP or ExternalIPPool changes, release the previous one first.
 		if err := c.releaseEgressIP(egress.Name, prevIP, prevIPPool); err != nil {
-			return nil, err
+			return nil, egress, err
 		}
 	}
 
 	// Skip allocating EgressIP if ExternalIPPool is not specified and return whatever user specifies.
 	if egress.Spec.ExternalIPPool == "" {
-		return net.ParseIP(egress.Spec.EgressIP), nil
+		return net.ParseIP(egress.Spec.EgressIP), egress, nil
 	}
 
 	if !c.externalIPAllocator.IPPoolExists(egress.Spec.ExternalIPPool) {
 		// The IP pool has been deleted, reclaim the IP from the Egress API.
 		if egress.Spec.EgressIP != "" {
-			if err := c.updateEgressIP(egress, ""); err != nil {
-				return nil, err
+			if updatedEgress, err := c.updateEgressIP(egress, ""); err != nil {
+				return nil, updatedEgress, err
```

Comment:
```suggestion
				return nil, egress, err
```
otherwise it still returns `nil` to caller which could panic.